File: references/java-sdk/src/main/java/blue/language/sdk/MyOsPermissions.java
package blue.language.sdk;

import blue.language.Blue;
import blue.language.model.Node;

import java.util.ArrayList;
import java.util.List;

public final class MyOsPermissions {

    private static final Blue BLUE = new Blue();

    public Boolean read;
    public Boolean write;
    public Boolean allOps;
    public List<String> singleOps;

    private MyOsPermissions() {
        this.singleOps = new ArrayList<String>();
    }

    public static MyOsPermissions create() {
        return new MyOsPermissions();
    }

    public MyOsPermissions read(boolean value) {
        this.read = value;
        return this;
    }

    public MyOsPermissions write(boolean value) {
        this.write = value;
        return this;
    }

    public MyOsPermissions allOps(boolean value) {
        this.allOps = value;
        return this;
    }

    public MyOsPermissions singleOps(String... operations) {
        this.singleOps = new ArrayList<String>();
        if (operations != null) {
            for (String operation : operations) {
                if (operation == null || operation.trim().isEmpty()) {
                    continue;
                }
                this.singleOps.add(operation.trim());
            }
        }
        return this;
    }

    public Node build() {
        return BLUE.objectToNode(this);
    }
}

File: libs/sdk-dsl/src/lib/myos-permissions.ts
import type { JsonObject } from './types';

export class MyOsPermissions {
  private readValue: boolean | undefined;
  private shareValue: boolean | undefined;
  private allOpsValue: boolean | undefined;
  private singleOpsValue: string[] = [];
  private singleOpsSet = false;

  static create(): MyOsPermissions {
    return new MyOsPermissions();
  }

  read(value: boolean): this {
    this.readValue = value;
    return this;
  }

  write(value: boolean): this {
    this.shareValue = value;
    return this;
  }

  allOps(value: boolean): this {
    this.allOpsValue = value;
    return this;
  }

  singleOps(...operations: Array<string | null | undefined>): this {
    this.singleOpsSet = true;
    this.singleOpsValue = operations
      .map((operation) => operation?.trim() ?? '')
      .filter((operation) => operation.length > 0);
    return this;
  }

  build(): JsonObject {
    const result: JsonObject = {};
    if (this.readValue !== undefined) {
      result.read = this.readValue;
    }
    if (this.shareValue !== undefined) {
      result.share = this.shareValue;
    }
    if (this.allOpsValue !== undefined) {
      result.allOps = this.allOpsValue;
    }
    if (this.singleOpsSet) {
      result.singleOps = [...this.singleOpsValue];
    }
    return result;
  }
}
